File: 121. Best Time to Buy and Sell Stock/maxProfit.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        if(n < 2) return 0;
        int result = 0;
        int bestBuy  = prices[0];
        int bestSell = prices[1];
        for(int i = 0; i < n; ++i){
            if(prices[i] < bestBuy){
                result = max(result, bestSell - bestBuy);
                bestBuy = prices[i];
                bestSell = prices[i];    
            }else if(prices[i] > bestSell){
                bestSell = prices[i];
            }
        }
        return result = max(result, bestSell - bestBuy);        
    }
};

File: 121. Best Time to Buy and Sell Stock/max_profit.kt
class Solution {
    fun maxProfit(prices: IntArray): Int {
        var buyDay = 0
        val lastDay = prices.size - 1
        var ans = 0
        var sellDay = 1
        while (buyDay < lastDay && sellDay < prices.size) {
            val current = prices[sellDay] - prices[buyDay]
            if (current > ans) {
                ans = current
            }
            if (prices[sellDay] < prices[buyDay]) {
                buyDay = sellDay
            }
            sellDay++
        }
        return ans
    }
}
